### PR TITLE
Evarsolve.recheck_applications: use Typing.checked_appvect

### LINKED ISSUE #19556: Insufficient use of template polymorphism in elaboration
~~~coq

Section Lists.

  Universe ua.

  Variable A : Type@{ua}.

  Constraint list.u0 < ua.

  Fail Definition tl (l:list A) :=
    match l with
      | nil => nil
      | cons a m => m
    end.
(* In environment
A : Type
l : list A
a : A
m : list A
The term "m" has type "list A" while it is expected to have type "list ?T@{l1:=(a :: m)%list}"
(unable to find a well-typed instantiation for "?T": cannot ensure that "Type@{ua}" is a subtype of
"Type@{foo.5}").
 *)

  (* with type annotation it works: *)
  Definition tl (l:list A) : list A :=
    match l with
    | nil => nil
    | cons a m => m
    end.
~~~

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #19556 (the adhoc implementation did not understand template poly)

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: pretyping/evarsolve.ml [hunk 1]
```diff
@@ -216,30 +216,19 @@ let add_conv_oriented_pb ?(tail=true) (pbty,env,t1,t2) evd =
...
 (* We retype applications to ensure the universe constraints are collected *)
-
 exception IllTypedInstance of env * evar_map * EConstr.types * EConstr.types
-exception IllTypedInstanceFun of env * evar_map * EConstr.constr * EConstr.types
+
+let checked_appvect, checked_appvect_hook = Hook.make ()
 
 let recheck_applications unify flags env evdref t =
   let rec aux env t =
+    (* the order matters: if the sub-applications are incorrect, checked_appvect may fail badly *)
```
Comment (SkySkimmer) [inline pretyping/evarsolve.ml:225]: both errors were due to this

**Comment by SkySkimmer** [comment]
@coqbot ci minimize ci-hott

**Comment by mrhaandi** [comment]
I tried the proposed fix, and I still have (not exactly the same, but similar) universe inconsistencies.
Here is an example `D.v`:
```coq
Structure myType := MyType { myType_X :> Type }.

Canonical Structure myType_CS X := MyType X.

Require Import ssreflect.

Lemma dummy (s' : (MyType (unit * unit))) : True.
Proof.
  move:s' => [? ?].
Admitted.

Axiom ax : forall (x : myType * unit), (fun t => match t with (_, _) => True end) x.
```
where `coqc D.v` results in
```
In environment
x : myType * unit
The term "x" has type "(myType * unit)%type"
while it is expected to have type "(?T * ?T0)%type"
(unable to find a well-typed instantiation for "?T": cannot ensure that
"Type@{myType.u0+1}" is a subtype of "Type@{D.10}").
```
Curiously, then I use `Alt+end` in vscode no error is produces, whereas stepwise `Alt+arrowdown` produces an error.

**Comment by SkySkimmer** [comment]
That seems to be a bug in Admitted, such that fixing #19556 was only a workaround for the first example.

**Comment by SkySkimmer** [comment]
@coqbot resume ci-minimize ci-hott https://gist.githubusercontent.com/SkySkimmer/549512ef6411c3931f0a52787b6e79ab/raw/c65cc8b1a9290a936933ff176dc87ba9cd1c10a1/tmp.v

**Comment by SkySkimmer** [comment]
@JasonGross it seems to have failed to make `merely` opaque, I guess the error message changed too much to be recognized?

@coqbot resume ci-minimize ci-hott https://gist.githubusercontent.com/SkySkimmer/549512ef6411c3931f0a52787b6e79ab/raw/42d1fdda824361e421f8440863c6353e94df3378/tmp.v

**Comment by mrhaandi** [comment]
Here is an example with three files (no `Admitted.`), for which `coqc A.v; coqc B.v; coqc C.v` in 8.20 works, but the current proposed PR fails:
`A.v`:
```coq
Structure myType := MyType { myType_X :> Type }.

Canonical Structure myType_CS X := MyType X.
```
`B.v`
```coq
Require Import A ssreflect.

Axiom A1 : (unit * myType) -> Prop.

Lemma dummy (s' : (MyType (unit * unit))) : True.
Proof.
  move:s' => [? ?]; exact I.
Qed.
```
`C.v`:
```coq
Require Import A.

Axiom f : myType -> unit -> Prop.
Axiom ax : forall x, (fun '(M, t) => f M t) x.

Require Import B.
```
with the error:
```
Universe inconsistency. Cannot enforce prod.u0 <= myType_CS.u0 because
myType_CS.u0 <= myType.u0 < prod.u0.
```

**Comment by SkySkimmer** [comment]
@mrhaandi please continue discussion of https://github.com/coq/coq/issues/19564 there since it seems not to be about this PR

File: pretyping/evarsolve.ml [hunk 2]
```diff
@@ -216,31 +216,29 @@ let add_conv_oriented_pb ?(tail=true) (pbty,env,t1,t2) evd =
...
+    | _ -> ()
+  in
+  try aux env t
+  with PretypeError (env,sigma,e) ->
+  match e with
+  | CantApplyBadTypeExplained (((_,expected,argty),_,_),_) ->
+    raise (IllTypedInstance (env,sigma,argty, expected))
+  | TypingError (CantApplyNonFunctional (fj,_)) ->
+    raise (IllTypedInstanceFun (env,sigma,fj.uj_val,fj.uj_type))
+  | _ -> assert false
```
Comment (ppedrot) [inline pretyping/evarsolve.ml:241]: Is it really impossible though? We call quite complicated code in check_appvect, including arbitrary calls to the retyper. Couldn't we just reraise the exception?

Comment (ppedrot) [reply pretyping/evarsolve.ml:241]: For instance it could fail in define_pure_evar_as_product with some universe error or something.

Comment (SkySkimmer) [reply pretyping/evarsolve.ml:241]: I think reraising the exn means it won't get handled properly
I'd rather have the assert false

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Possibly some small changes in either direction, I think we should accept them in the name of being correct.

**Comment by ppedrot** [comment]
I'm side-stepping @mattam82 to @coqbot merge now

**Pull request merged**